clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/view/GoogleApplicationProvider.java

```diff
@@ -0,0 +1,115 @@
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableSet;
+import com.netflix.spinnaker.cats.cache.Cache;
+import com.netflix.spinnaker.cats.cache.CacheData;
+import com.netflix.spinnaker.cats.cache.RelationshipCacheFilter;
+import com.netflix.spinnaker.clouddriver.google.GoogleCloudProvider;
+import com.netflix.spinnaker.clouddriver.google.cache.Keys;
+import com.netflix.spinnaker.clouddriver.google.cache.Keys.Namespace;
+import com.netflix.spinnaker.clouddriver.google.model.GoogleApplication;
+import com.netflix.spinnaker.clouddriver.model.Application;
+import com.netflix.spinnaker.clouddriver.model.ApplicationProvider;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+final class GoogleApplicationProvider implements ApplicationProvider {
+
+  private final Cache cacheView;
+  private final ObjectMapper objectMapper;
+
+  @Autowired
+  GoogleApplicationProvider(Cache cacheView, ObjectMapper objectMapper) {
+    this.cacheView = cacheView;
+    this.objectMapper = objectMapper;
+  }
+
+  @Override
+  public Set<? extends Application> getApplications(boolean expand) {
+
+    RelationshipCacheFilter filter =
+        expand ? RelationshipCacheFilter.include(CLUSTERS.getNs()) : RelationshipCacheFilter.none();
+    Collection<CacheData> data =
+        cacheView.getAll(
+            APPLICATIONS.getNs(),
+            cacheView.filterIdentifiers(APPLICATIONS.getNs(), GoogleCloudProvider.getID() + ":*"),
+            filter);
+    return data.stream().map(this::applicationFromCacheData).collect(toSet());
+  }
+
+  @Override
+  public Application getApplication(String name) {
+
+    CacheData cacheData =
+        cacheView.get(
+            APPLICATIONS.getNs(),
+            Keys.getApplicationKey(name),
+            RelationshipCacheFilter.include(CLUSTERS.getNs(), INSTANCES.getNs()));
+    if (cacheData == null) {
+      return null;
+    }
+
+    return applicationFromCacheData(cacheData);
+  }
+
+  private GoogleApplication.View applicationFromCacheData(CacheData cacheData) {
+
+    GoogleApplication application =
+        objectMapper.convertValue(cacheData.getAttributes(), GoogleApplication.class);
+    if (application == null) {
+      return null;
+    }
+
+    GoogleApplication.View applicationView = application.getView();
+
+    Collection<String> clusters = getRelationships(cacheData, CLUSTERS);
+    clusters.forEach(
+        key -> {
+          Map<String, String> parsedKey = Keys.parse(key);
+          applicationView
+              .getClusterNames()
+              .get(parsedKey.get("account"))
+              .add(parsedKey.get("name"));
```

Comment:
Does the `withDefault` on the `GoogleApplication.clusterNames` work even when called from Java or do we need to handle the case where `.get(parsedKey.get("account"))` is null?

Reply:
Yeah, it works. I wanted to convert GoogleApplication to Java too, and turn clusterNames into a SetMultimap, which makes this simpler... but I was worried I'd break the serialization.